### PR TITLE
Attack Detector - Refactoring

### PR DESCRIPTION
**Pull request by Vxatz**

### CODE REVIEW

File: alert-combiner-py/src/dynamo_utils.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
+import logging
+from datetime import datetime
+import time
+from src.constants import ALERTS_LOOKBACK_WINDOW_IN_HOURS
+import pandas as pd
+from src.utils import Utils 
+
+item_id_prefix = ""
+
+class DynamoUtils:
```
Comment (christian-forta) [inline alert-combiner-py/src/dynamo_utils.py:10]: seems like a lot of copy paste; can we refactor to make this a bit more modular?

File: alert-combiner-py/src/utils.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        """
+        this function determines whether address is a valid address
+        :return: is_address: bool
+        """
+        if addresses is None:
+            return True
+
+        is_address = True
+        for address in addresses.split(','):
+            for c in ['a', 'b', 'c', 'd', 'e', 'f', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9']:
```
Comment (christian-forta) [inline alert-combiner-py/src/utils.py:43]: this is my super ugly code; do you know how to better check this?

File: alert-combiner-py/src/agent.py [hunk 3]
```diff
@@ -423,23 +351,28 @@ def detect_attack(w3, alert_event: forta_agent.alert_event.AlertEvent) -> list:
...
+                            new_alert_data = pd.DataFrame([[stage, datetime.strptime(alert_event.alert.created_at[:-4] + 'Z', "%Y-%m-%dT%H:%M:%S.%fZ"), alert_anomaly_score, alert_event.alert_hash, alert_event.bot_id, alert_event.alert.alert_id, alert_event.alert.addresses, alert_event.alert.source.transaction_hash]], columns=['stage', 'created_at', 'anomaly_score', 'alert_hash', 'bot_id', 'alert_id', 'addresses', 'transaction_hash'])
+                        alert_data_cluster = pd.concat([alert_data_cluster, new_alert_data], ignore_index=True, axis=0)
+                        logging.info(f"alert {alert_event.alert_hash} - alert data size for cluster {cluster} now: {len(alert_data_cluster)}")
 
                         # add new alert and purge old alerts
-                        ALERT_DATA[cluster] = alert_data[alert_data['created_at'] > start_date]
-                        alert_data = ALERT_DATA[cluster]
-                        logging.info(f"alert {alert_event.alert_hash} - alert data size for cluster {cluster} now (after date pruning): {len(alert_data)}")
+                        alert_data_cluster = alert_data_cluster[alert_data_cluster['created_at'] > start_date]
```
Comment (christian-forta) [inline alert-combiner-py/src/agent.py:372]: do we need this as dynamo does auto expiry?

**Review by christian-forta — APPROVED** [review]
looks good vasilis. but lets see how it does on the unit tests :)

File: alert-combiner-py/src/utils.py (hunk 2 above)
Comment (apehex) [reply alert-combiner-py/src/utils.py:43]: I think you can go for:

```python
_test = any([9*c in address.lower() for c in 'abcdef0123456789'])
if _test: pass
```

I fell for the trap of the regex too, it's 8 repetitions! :sweat_smile: 

```python
_r = re.compile(r'([a-f0-9])\1{8}')
```

File: alert-combiner-py/src/dynamo_utils.py (hunk 1 above)
Comment (Vxatz) [reply alert-combiner-py/src/dynamo_utils.py:11]: @christian-forta, I refactored the code a bit, but besides the read/write for the fp mitigation and end user attacks clusters, they all have some differences and I couldn't find a way of creating a more generic function without it being too "ugly".
If you have any ideas, I can update it.

**Review by christian-forta — APPROVED** [review]
awesome. thank you, Vasilis!

**Comment by christian-forta** [comment]
Ill deploy it to beta now from this branch.

**Pull request merged**